File: modules/spv-mobile/ios/SpvMobileModule.swift
import ExpoModulesCore

// MARK: - Record types (mirror Kotlin SpvMobileModule records exactly)

struct UtxoRecord: Record {
  init() {}
  @Field var txid: String = ""
  @Field var index: Int = 0
  @Field var value: Double = 0          // u64 → Double (JS-safe up to 2^53)
  @Field var height: Double = 0
  @Field var is_coinbase: Bool = false
  @Field var script_pubkey_hex: String = ""
}

struct BalanceInfoRecord: Record {
  init() {}
  @Field var confirmed: Double = 0
  @Field var utxo_count: Int = 0
  @Field var height: Double = 0
}

struct NodeStatusRecord: Record {
  init() {}
  @Field var height: Double = 0
  @Field var peer_count: Int = 0
  @Field var tip_hash: String = ""
  @Field var anchor_loaded: Bool = false
}

struct TxRecordRecord: Record {
  init() {}
  @Field var txid: String = ""
  @Field var height: Double = 0
  @Field var net_sats: Double = 0       // i64 → Double
  @Field var direction: String = "in"
}

struct HtlcInfoRecord: Record {
  init() {}
  @Field var secret_hash_hex: String = ""
  @Field var recipient_address: String = ""
  @Field var refund_address: String = ""
  @Field var timeout_height: Double = 0
}

struct WifKeyRecord: Record {
  init() {}
  @Field var address: String = ""
  @Field var pubkey_hex: String = ""
}

struct LightClientConfigRecord: Record {
  init() {}
  @Field var seedlist_path: String = ""
  @Field var extra_peer: String? = nil
  @Field var start_height: Double = 0
  @Field var start_hash: String? = nil
}

struct AgreementParamsRecord: Record {
  init() {}
  @Field var template_type: String = ""
  @Field var payer_address: String = ""
  @Field var payee_address: String = ""
  @Field var total_amount_sats: Double = 0
  @Field var timeout_height: Double = 0
  @Field var secret_hash_hex: String = ""
  @Field var asset_reference: String? = nil
  @Field var payment_reference: String? = nil
  @Field var document_hash: String? = nil
}

struct HtlcClaimParamsRecord: Record {
  init() {}
  @Field var htlc_txid: String = ""
  @Field var htlc_index: Int = 0
  @Field var htlc_value: Double = 0
  @Field var htlc_script_hex: String = ""
  @Field var preimage_hex: String = ""
  @Field var seed_hex: String = ""
  @Field var key_index: Int = 0
  @Field var to_address: String = ""
  @Field var fee_sats: Double = 0
}

struct HtlcRefundParamsRecord: Record {
  init() {}
  @Field var htlc_txid: String = ""
  @Field var htlc_index: Int = 0
  @Field var htlc_value: Double = 0
  @Field var htlc_script_hex: String = ""
  @Field var seed_hex: String = ""
  @Field var key_index: Int = 0
  @Field var to_address: String = ""
  @Field var fee_sats: Double = 0
  @Field var timeout_height: Double = 0
}

// MARK: - Helpers

private func requireOk(_ result: String) throws -> String {
  if result.hasPrefix("ERROR:") {
    throw NSError(domain: "SpvMobile", code: 1,
      userInfo: [NSLocalizedDescriptionKey: String(result.dropFirst(6))])
  }
  return result
}

private func utxosToJson(_ utxos: [UtxoRecord]) throws -> String {
  let arr = utxos.map { u -> [String: Any] in
    return [
      "txid": u.txid,
      "index": u.index,
      "value": Int64(u.value),
      "height": Int64(u.height),
      "is_coinbase": u.is_coinbase,
      "script_pubkey_hex": u.script_pubkey_hex,
    ]
  }
  let data = try JSONSerialization.data(withJSONObject: arr)
  return String(data: data, encoding: .utf8) ?? "[]"
}

private func parseJson(_ json: String) throws -> [String: Any] {
  guard let data = json.data(using: .utf8),
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
  else { throw NSError(domain: "SpvMobile", code: 2, userInfo: [NSLocalizedDescriptionKey: "JSON parse error"]) }
  return obj
}

private func parseJsonArray(_ json: String) throws -> [[String: Any]] {
  guard let data = json.data(using: .utf8),
        let arr = try JSONSerialization.jsonObject(with: data) as? [[String: Any]]
  else { throw NSError(domain: "SpvMobile", code: 2, userInfo: [NSLocalizedDescriptionKey: "JSON array parse error"]) }
  return arr
}

// MARK: - Module
//
// This module calls UniFFI-generated Swift functions from spv_mobile.swift,
// which is placed alongside this file in ios/ and compiled into the same pod.
// Run `uniffi-bindgen generate rust-bridge/src/spv_mobile.udl --language swift
//      --out-dir modules/spv-mobile/ios/` on a Mac to produce spv_mobile.swift.

public class SpvMobileModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SpvMobileModule")

    // MARK: Wallet

    AsyncFunction("generateMnemonic") { () throws -> String in
      try requireOk(nativeGenerateMnemonic())
    }

    AsyncFunction("mnemonicToSeed") { (mnemonic: String, passphrase: String) throws -> String in
      try requireOk(nativeMnemonicToSeed(mnemonic: mnemonic, passphrase: passphrase))
    }

    AsyncFunction("deriveAddress") { (seedHex: String, index: Int) throws -> String in
      try requireOk(nativeDeriveAddress(seedHex: seedHex, index: UInt32(index)))
    }

    AsyncFunction("derivePrivkeyHex") { (seedHex: String, index: Int) throws -> String in
      try requireOk(nativeDerivePrivkeyHex(seedHex: seedHex, index: UInt32(index)))
    }

    Function("validateAddress") { (address: String) -> Bool in
      nativeValidateAddress(address: address)
    }

    AsyncFunction("exportWif") { (seedHex: String, index: Int) throws -> String in
      try requireOk(nativeExportWif(seedHex: seedHex, index: UInt32(index)))
    }

    AsyncFunction("importWif") { (wif: String) throws -> WifKeyRecord in
      let json = try requireOk(nativeImportWif(wif: wif))
      let obj  = try parseJson(json)
      var r    = WifKeyRecord()
      r.address    = obj["address"]    as? String ?? ""
      r.pubkey_hex = obj["pubkey_hex"] as? String ?? ""
      return r
    }

    // MARK: Transaction building

    AsyncFunction("buildSendTx") { (utxos: [UtxoRecord], toAddress: String,
                                    amountSats: Double, feeSats: Double,
                                    seedHex: String, fromIndex: Int) throws -> String in
      let utxosJson = try utxosToJson(utxos)
      return try requireOk(nativeBuildSendTx(
        utxosJson: utxosJson, toAddress: toAddress,
        amountSats: UInt64(amountSats), feeSats: UInt64(feeSats),
        seedHex: seedHex, fromIndex: UInt32(fromIndex)
      ))
    }

    // MARK: HTLC

    AsyncFunction("encodeHtlcv1") { (secretHashHex: String, recipientAddress: String,
                                     refundAddress: String, timeoutHeight: Double) throws -> String in
      try requireOk(nativeEncodeHtlcv1(
        secretHashHex: secretHashHex, recipientAddress: recipientAddress,
        refundAddress: refundAddress, timeoutHeight: UInt64(timeoutHeight)
      ))
    }

    AsyncFunction("decodeHtlcv1") { (scriptHex: String) throws -> HtlcInfoRecord in
      let json = try requireOk(nativeDecodeHtlcv1(scriptHex: scriptHex))
      let obj  = try parseJson(json)
      var r    = HtlcInfoRecord()
      r.secret_hash_hex    = obj["secret_hash_hex"]    as? String ?? ""
      r.recipient_address  = obj["recipient_address"]  as? String ?? ""
      r.refund_address     = obj["refund_address"]     as? String ?? ""
      r.timeout_height     = obj["timeout_height"].flatMap { Double(exactly: $0 as AnyObject) } ?? 0
      return r
    }

    AsyncFunction("buildHtlcClaimTx") { (params: HtlcClaimParamsRecord) throws -> String in
      try requireOk(nativeBuildHtlcClaimTx(
        htlcTxid: params.htlc_txid, htlcIndex: UInt32(params.htlc_index),
        htlcValue: UInt64(params.htlc_value), htlcScriptHex: params.htlc_script_hex,
        preimageHex: params.preimage_hex, seedHex: params.seed_hex,
        keyIndex: UInt32(params.key_index), toAddress: params.to_address,
        feeSats: UInt64(params.fee_sats)
      ))
    }

    AsyncFunction("buildHtlcRefundTx") { (params: HtlcRefundParamsRecord) throws -> String in
      try requireOk(nativeBuildHtlcRefundTx(
        htlcTxid: params.htlc_txid, htlcIndex: UInt32(params.htlc_index),
        htlcValue: UInt64(params.htlc_value), htlcScriptHex: params.htlc_script_hex,
        seedHex: params.seed_hex, keyIndex: UInt32(params.key_index),
        toAddress: params.to_address, feeSats: UInt64(params.fee_sats),
        timeoutHeight: UInt64(params.timeout_height)
      ))
    }

    // MARK: Agreement

    AsyncFunction("computeAgreementHash") { (agreementJson: String) throws -> String in
      try requireOk(nativeComputeAgreementHash(agreementJson: agreementJson))
    }

    AsyncFunction("createAgreement") { (params: AgreementParamsRecord) throws -> String in
      try requireOk(nativeCreateAgreement(
        templateType: params.template_type,
        payerAddress: params.payer_address,
        payeeAddress: params.payee_address,
        totalAmountSats: UInt64(params.total_amount_sats),
        timeoutHeight: UInt64(params.timeout_height),
        secretHashHex: params.secret_hash_hex,
        assetReference: params.asset_reference ?? "",
        paymentReference: params.payment_reference ?? "",
        documentHash: params.document_hash ?? ""
      ))
    }

    // MARK: SPV

    AsyncFunction("verifyMerkleProof") { (txidHex: String, merkleRootHex: String,
                                          proofHexNodes: [String], index: Double) throws -> Bool in
      let proofData = try JSONSerialization.data(withJSONObject: proofHexNodes)
      let proofJson = String(data: proofData, encoding: .utf8) ?? "[]"
      return nativeVerifyMerkleProof(
        txidHex: txidHex, merkleRootHex: merkleRootHex,
        proofHexNodesJson: proofJson, index: UInt64(index)
      )
    }

    Function("getSyncedHeight") { () -> Double in Double(nativeGetSyncedHeight()) }
    Function("getTipHash") { () -> String in nativeGetTipHash() }

    // MARK: RPC

    AsyncFunction("rpcGetStatus") { (rpcUrl: String, authToken: String?) throws -> NodeStatusRecord in
      let json = try requireOk(nativeRpcGetStatus(rpcUrl: rpcUrl, authToken: authToken ?? ""))
      let obj  = try parseJson(json)
      var r    = NodeStatusRecord()
      r.height        = (obj["height"]       as? Double) ?? 0
      r.peer_count    = (obj["peer_count"]   as? Int)    ?? 0
      r.tip_hash      = (obj["tip_hash"]     as? String) ?? ""
      r.anchor_loaded = (obj["anchor_loaded"] as? Bool)  ?? false
      return r
    }

    AsyncFunction("rpcGetBalance") { (rpcUrl: String, authToken: String?, address: String) throws -> BalanceInfoRecord in
      let json = try requireOk(nativeRpcGetBalance(rpcUrl: rpcUrl, authToken: authToken ?? "", address: address))
      let obj  = try parseJson(json)
      var r    = BalanceInfoRecord()
      r.confirmed   = (obj["confirmed"]   as? Double) ?? 0
      r.utxo_count  = (obj["utxo_count"]  as? Int)    ?? 0
      r.height      = (obj["height"]      as? Double) ?? 0
      return r
    }

    AsyncFunction("rpcGetUtxos") { (rpcUrl: String, authToken: String?, address: String) throws -> [UtxoRecord] in
      let json = try requireOk(nativeRpcGetUtxos(rpcUrl: rpcUrl, authToken: authToken ?? "", address: address))
      return try parseJsonArray(json).map { obj in
        var r = UtxoRecord()
        r.txid             = (obj["txid"]            as? String) ?? ""
        r.index            = (obj["index"]           as? Int)    ?? 0
        r.value            = (obj["value"]           as? Double) ?? 0
        r.height           = (obj["height"]          as? Double) ?? 0
        r.is_coinbase      = (obj["is_coinbase"]     as? Bool)   ?? false
        r.script_pubkey_hex = (obj["script_pubkey_hex"] as? String) ?? ""
        return r
      }
    }

    AsyncFunction("rpcGetHistory") { (rpcUrl: String, authToken: String?, address: String) throws -> [TxRecordRecord] in
      let json = try requireOk(nativeRpcGetHistory(rpcUrl: rpcUrl, authToken: authToken ?? "", address: address))
      return try parseJsonArray(json).map { obj in
        var r = TxRecordRecord()
        r.txid      = (obj["txid"]      as? String) ?? ""
        r.height    = (obj["height"]    as? Double) ?? 0
        r.net_sats  = (obj["net_sats"]  as? Double) ?? 0
        r.direction = (obj["direction"] as? String) ?? "in"
        return r
      }
    }

    AsyncFunction("rpcGetFeeRate") { (rpcUrl: String, authToken: String?) throws -> Double in
      Double(nativeRpcGetFeeRate(rpcUrl: rpcUrl, authToken: authToken ?? ""))
    }

    AsyncFunction("rpcSubmitTx") { (rpcUrl: String, authToken: String?, txHex: String) throws -> String in
      try requireOk(nativeRpcSubmitTx(rpcUrl: rpcUrl, authToken: authToken ?? "", txHex: txHex))
    }

    // MARK: Light client

    AsyncFunction("startLightClient") { (config: LightClientConfigRecord) throws in
      let resolved = self.resolveSeedlistPath(config.seedlist_path)
      let result = nativeStartLightClient(
        seedlistPath: resolved,
        extraPeer: config.extra_peer ?? "",
        startHeight: UInt64(config.start_height),
        startHash: config.start_hash ?? ""
      )
      _ = try requireOk(result)
    }

    Function("stopLightClient") { () in nativeStopLightClient() }
    Function("isSyncing") { () -> Bool in nativeIsSyncing() }
    Function("peerCount") { () -> Int in Int(nativePeerCount()) }

    AsyncFunction("broadcastTx") { (txHex: String) throws -> String in
      try requireOk(nativeBroadcastTx(txHex: txHex))
    }
  }

  // If path starts with "assets/", copy the bundled file to the app's Documents
  // directory (writable by Rust's std::fs) and return that path.
  private func resolveSeedlistPath(_ path: String) -> String {
    guard path.hasPrefix("assets/") else { return path }
    let assetName = String(path.dropFirst(7))
    guard let bundlePath = Bundle.main.path(forResource: assetName, ofType: nil) else { return path }
    let dest = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("seedlist.txt")
    try? FileManager.default.copyItem(atPath: bundlePath, toPath: dest.path)
    return dest.path
  }
}

File: modules/spv-mobile/android/src/main/java/expo/modules/spvmobile/SpvMobileModule.kt
package expo.modules.spvmobile

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

// ---------- Record types (match UDL dictionaries exactly) ----------

class UtxoRecord : Record {
    @Field var txid: String = ""
    @Field var index: Int = 0
    @Field var value: Long = 0
    @Field var height: Long = 0
    @Field var is_coinbase: Boolean = false
    @Field var script_pubkey_hex: String = ""
}

class BalanceInfoRecord : Record {
    @Field var confirmed: Long = 0
    @Field var utxo_count: Int = 0
    @Field var height: Long = 0
}

class NodeStatusRecord : Record {
    @Field var height: Long = 0
    @Field var peer_count: Int = 0
    @Field var tip_hash: String = ""
    @Field var anchor_loaded: Boolean = false
}

class TxRecordRecord : Record {
    @Field var txid: String = ""
    @Field var height: Long = 0
    @Field var net_sats: Long = 0
    @Field var direction: String = "in"
}

class HtlcInfoRecord : Record {
    @Field var secret_hash_hex: String = ""
    @Field var recipient_address: String = ""
    @Field var refund_address: String = ""
    @Field var timeout_height: Long = 0
}

class WifKeyRecord : Record {
    @Field var address: String = ""
    @Field var pubkey_hex: String = ""
}

class LightClientConfigRecord : Record {
    @Field var seedlist_path: String = ""
    @Field var extra_peer: String? = null
    @Field var start_height: Long = 0
    @Field var start_hash: String? = null
}

class AgreementParamsRecord : Record {
    @Field var template_type: String = ""
    @Field var payer_address: String = ""
    @Field var payee_address: String = ""
    @Field var total_amount_sats: Long = 0
    @Field var timeout_height: Long = 0
    @Field var secret_hash_hex: String = ""
    @Field var asset_reference: String? = null
    @Field var payment_reference: String? = null
    @Field var document_hash: String? = null
}

// Expo DSL caps at 8 typed params; wrap 9-param HTLC functions in a Record.
class HtlcClaimParamsRecord : Record {
    @Field var htlc_txid: String = ""
    @Field var htlc_index: Int = 0
    @Field var htlc_value: Long = 0
    @Field var htlc_script_hex: String = ""
    @Field var preimage_hex: String = ""
    @Field var seed_hex: String = ""
    @Field var key_index: Int = 0
    @Field var to_address: String = ""
    @Field var fee_sats: Long = 0
}

class HtlcRefundParamsRecord : Record {
    @Field var htlc_txid: String = ""
    @Field var htlc_index: Int = 0
    @Field var htlc_value: Long = 0
    @Field var htlc_script_hex: String = ""
    @Field var seed_hex: String = ""
    @Field var key_index: Int = 0
    @Field var to_address: String = ""
    @Field var fee_sats: Long = 0
    @Field var timeout_height: Long = 0
}

// ---------- JNI singleton ----------

internal object SpvNative {
    init {
        System.loadLibrary("spv_mobile")
    }

    external fun nativePeerCount(): Int
    external fun nativeGetSyncedHeight(): Long
    external fun nativeIsSyncing(): Byte
    external fun nativeGetTipHash(): String
    external fun nativeStartLightClient(
        seedlistPath: String, extraPeer: String, startHeight: Long, startHash: String
    ): String
    external fun nativeStopLightClient()
    external fun nativeBroadcastTx(txHex: String): String

    external fun nativeGenerateMnemonic(): String
    external fun nativeMnemonicToSeed(mnemonic: String, passphrase: String): String
    external fun nativeDeriveAddress(seedHex: String, index: Int): String
    external fun nativeDerivePrivkeyHex(seedHex: String, index: Int): String
    external fun nativeValidateAddress(address: String): Byte
    external fun nativeExportWif(seedHex: String, index: Int): String
    external fun nativeImportWif(wif: String): String

    external fun nativeBuildSendTx(
        utxosJson: String, toAddress: String,
        amountSats: Long, feeSats: Long, seedHex: String, fromIndex: Int
    ): String

    external fun nativeEncodeHtlcv1(
        secretHashHex: String, recipientAddress: String,
        refundAddress: String, timeoutHeight: Long
    ): String
    external fun nativeDecodeHtlcv1(scriptHex: String): String
    external fun nativeBuildHtlcClaimTx(
        htlcTxid: String, htlcIndex: Int, htlcValue: Long, htlcScriptHex: String,
        preimageHex: String, seedHex: String, keyIndex: Int, toAddress: String, feeSats: Long
    ): String
    external fun nativeBuildHtlcRefundTx(
        htlcTxid: String, htlcIndex: Int, htlcValue: Long, htlcScriptHex: String,
        seedHex: String, keyIndex: Int, toAddress: String, feeSats: Long, timeoutHeight: Long
    ): String

    external fun nativeComputeAgreementHash(agreementJson: String): String
    external fun nativeCreateAgreement(
        templateType: String, payerAddress: String, payeeAddress: String,
        totalAmountSats: Long, timeoutHeight: Long, secretHashHex: String,
        assetReference: String, paymentReference: String, documentHash: String
    ): String

    external fun nativeVerifyMerkleProof(
        txidHex: String, merkleRootHex: String, proofHexNodesJson: String, index: Long
    ): Byte

    external fun nativeRpcGetStatus(rpcUrl: String, authToken: String): String
    external fun nativeRpcGetBalance(rpcUrl: String, authToken: String, address: String): String
    external fun nativeRpcGetUtxos(rpcUrl: String, authToken: String, address: String): String
    external fun nativeRpcGetHistory(rpcUrl: String, authToken: String, address: String): String
    external fun nativeRpcGetFeeRate(rpcUrl: String, authToken: String): Long
    external fun nativeRpcSubmitTx(rpcUrl: String, authToken: String, txHex: String): String
}

// ---------- Helpers ----------

private fun String.requireOk(): String {
    if (startsWith("ERROR:")) throw Exception(removePrefix("ERROR:"))
    return this
}

private fun List<UtxoRecord>.toJson(): String {
    val arr = JSONArray()
    forEach { u ->
        arr.put(JSONObject().apply {
            put("txid", u.txid)
            put("index", u.index)
            put("value", u.value)
            put("height", u.height)
            put("is_coinbase", u.is_coinbase)
            put("script_pubkey_hex", u.script_pubkey_hex)
        })
    }
    return arr.toString()
}

// ---------- Module ----------

class SpvMobileModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("SpvMobileModule")

        // --- Wallet ---

        AsyncFunction("generateMnemonic") {
            SpvNative.nativeGenerateMnemonic().requireOk()
        }

        AsyncFunction("mnemonicToSeed") { mnemonic: String, passphrase: String ->
            SpvNative.nativeMnemonicToSeed(mnemonic, passphrase).requireOk()
        }

        AsyncFunction("deriveAddress") { seedHex: String, index: Int ->
            SpvNative.nativeDeriveAddress(seedHex, index).requireOk()
        }

        AsyncFunction("derivePrivkeyHex") { seedHex: String, index: Int ->
            SpvNative.nativeDerivePrivkeyHex(seedHex, index).requireOk()
        }

        Function("validateAddress") { address: String ->
            SpvNative.nativeValidateAddress(address).toInt() == 1
        }

        AsyncFunction("exportWif") { seedHex: String, index: Int ->
            SpvNative.nativeExportWif(seedHex, index).requireOk()
        }

        AsyncFunction("importWif") { wif: String ->
            val json = SpvNative.nativeImportWif(wif).requireOk()
            val obj = JSONObject(json)
            val r = WifKeyRecord()
            r.address = obj.getString("address")
            r.pubkey_hex = obj.getString("pubkey_hex")
            r
        }

        // --- Transaction building ---

        AsyncFunction("buildSendTx") { utxos: List<UtxoRecord>, toAddress: String,
                                       amountSats: Long, feeSats: Long,
                                       seedHex: String, fromIndex: Int ->
            SpvNative.nativeBuildSendTx(utxos.toJson(), toAddress, amountSats, feeSats, seedHex, fromIndex)
                .requireOk()
        }

        // --- HTLC ---

        AsyncFunction("encodeHtlcv1") { secretHashHex: String, recipientAddress: String,
                                        refundAddress: String, timeoutHeight: Long ->
            SpvNative.nativeEncodeHtlcv1(secretHashHex, recipientAddress, refundAddress, timeoutHeight)
                .requireOk()
        }

        AsyncFunction("decodeHtlcv1") { scriptHex: String ->
            val json = SpvNative.nativeDecodeHtlcv1(scriptHex).requireOk()
            val obj = JSONObject(json)
            val r = HtlcInfoRecord()
            r.secret_hash_hex = obj.getString("secret_hash_hex")
            r.recipient_address = obj.getString("recipient_address")
            r.refund_address = obj.getString("refund_address")
            r.timeout_height = obj.getLong("timeout_height")
            r
        }

        AsyncFunction("buildHtlcClaimTx") { params: HtlcClaimParamsRecord ->
            SpvNative.nativeBuildHtlcClaimTx(
                params.htlc_txid, params.htlc_index, params.htlc_value,
                params.htlc_script_hex, params.preimage_hex, params.seed_hex,
                params.key_index, params.to_address, params.fee_sats
            ).requireOk()
        }

        AsyncFunction("buildHtlcRefundTx") { params: HtlcRefundParamsRecord ->
            SpvNative.nativeBuildHtlcRefundTx(
                params.htlc_txid, params.htlc_index, params.htlc_value,
                params.htlc_script_hex, params.seed_hex, params.key_index,
                params.to_address, params.fee_sats, params.timeout_height
            ).requireOk()
        }

        // --- Agreement ---

        AsyncFunction("computeAgreementHash") { agreementJson: String ->
            SpvNative.nativeComputeAgreementHash(agreementJson).requireOk()
        }

        AsyncFunction("createAgreement") { params: AgreementParamsRecord ->
            SpvNative.nativeCreateAgreement(
                params.template_type, params.payer_address, params.payee_address,
                params.total_amount_sats, params.timeout_height, params.secret_hash_hex,
                params.asset_reference ?: "", params.payment_reference ?: "",
                params.document_hash ?: ""
            ).requireOk()
        }

        // --- SPV ---

        AsyncFunction("verifyMerkleProof") { txidHex: String, merkleRootHex: String,
                                             proofHexNodes: List<String>, index: Long ->
            val proofJson = JSONArray().apply { proofHexNodes.forEach { put(it) } }.toString()
            val result = SpvNative.nativeVerifyMerkleProof(txidHex, merkleRootHex, proofJson, index)
            result.toInt() == 1
        }

        Function("getSyncedHeight") { -> SpvNative.nativeGetSyncedHeight() }
        Function("getTipHash") { -> SpvNative.nativeGetTipHash() }

        // --- RPC ---

        AsyncFunction("rpcGetStatus") { rpcUrl: String, authToken: String? ->
            val json = SpvNative.nativeRpcGetStatus(rpcUrl, authToken ?: "").requireOk()
            val obj = JSONObject(json)
            val r = NodeStatusRecord()
            r.height = obj.getLong("height")
            r.peer_count = obj.getInt("peer_count")
            r.tip_hash = obj.getString("tip_hash")
            r.anchor_loaded = obj.getBoolean("anchor_loaded")
            r
        }

        AsyncFunction("rpcGetBalance") { rpcUrl: String, authToken: String?, address: String ->
            val json = SpvNative.nativeRpcGetBalance(rpcUrl, authToken ?: "", address).requireOk()
            val obj = JSONObject(json)
            val r = BalanceInfoRecord()
            r.confirmed = obj.getLong("confirmed")
            r.utxo_count = obj.getInt("utxo_count")
            r.height = obj.getLong("height")
            r
        }

        AsyncFunction("rpcGetUtxos") { rpcUrl: String, authToken: String?, address: String ->
            val json = SpvNative.nativeRpcGetUtxos(rpcUrl, authToken ?: "", address).requireOk()
            val arr = JSONArray(json)
            val list = mutableListOf<UtxoRecord>()
            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                val r = UtxoRecord()
                r.txid = obj.getString("txid")
                r.index = obj.getInt("index")
                r.value = obj.getLong("value")
                r.height = obj.getLong("height")
                r.is_coinbase = obj.getBoolean("is_coinbase")
                r.script_pubkey_hex = obj.getString("script_pubkey_hex")
                list.add(r)
            }
            list
        }

        AsyncFunction("rpcGetHistory") { rpcUrl: String, authToken: String?, address: String ->
            val json = SpvNative.nativeRpcGetHistory(rpcUrl, authToken ?: "", address).requireOk()
            val arr = JSONArray(json)
            val list = mutableListOf<TxRecordRecord>()
            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                val r = TxRecordRecord()
                r.txid = obj.getString("txid")
                r.height = obj.getLong("height")
                r.net_sats = obj.getLong("net_sats")
                r.direction = obj.getString("direction")
                list.add(r)
            }
            list
        }

        AsyncFunction("rpcGetFeeRate") { rpcUrl: String, authToken: String? ->
            SpvNative.nativeRpcGetFeeRate(rpcUrl, authToken ?: "")
        }

        AsyncFunction("rpcSubmitTx") { rpcUrl: String, authToken: String?, txHex: String ->
            SpvNative.nativeRpcSubmitTx(rpcUrl, authToken ?: "", txHex).requireOk()
        }

        // --- Light client ---

        AsyncFunction("startLightClient") { config: LightClientConfigRecord ->
            val resolvedPath = resolveSeedlistPath(config.seedlist_path)
            SpvNative.nativeStartLightClient(
                resolvedPath, config.extra_peer ?: "",
                config.start_height, config.start_hash ?: ""
            ).requireOk()
        }

        Function("stopLightClient") { -> SpvNative.nativeStopLightClient() }
        Function("isSyncing") { -> SpvNative.nativeIsSyncing().toInt() == 1 }
        Function("peerCount") { -> SpvNative.nativePeerCount() }

        AsyncFunction("broadcastTx") { txHex: String ->
            SpvNative.nativeBroadcastTx(txHex).requireOk()
        }
    }

    // If the caller passes an "assets/..." path, copy the bundled asset to the
    // app's internal files dir (writable by Rust's std::fs) and return that path.
    private fun resolveSeedlistPath(path: String): String {
        if (!path.startsWith("assets/")) return path
        val assetName = path.removePrefix("assets/")
        val ctx = appContext.reactContext ?: return path
        val dest = File(ctx.filesDir, "seedlist.txt")
        try {
            ctx.assets.open(assetName).use { input ->
                dest.outputStream().use { out -> input.copyTo(out) }
            }
        } catch (_: Exception) { /* asset missing — Rust will report the error */ }
        return dest.absolutePath
    }
}
